server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
         }
     }
 
+     private static Query tryRewriteNumericLongSort(SearchContext searchContext, IndexReader reader) throws IOException {
+        // check sort
+        Sort sort = searchContext.sort().sort;
+        if (Sort.RELEVANCE.equals(sort)) return null;
+        if (Sort.INDEXORDER.equals(sort)) return null;
+        if (sort.getSort().length > 1) return null; // we need only a single sort
+        SortField sortField = sort.getSort()[0];
+
+        // check if this is a field of type Long, that is indexed and has doc values
+        String fieldName = sortField.getField();
+        final MappedFieldType fieldType = searchContext.mapperService().fullName(fieldName);
+        if (fieldType == null) return  null;
+        if (fieldType.typeName() != "long") return null;
```

Comment:
Can you use `.equals` instead? I don't we would get an interned string all the time, so it shouldn't really matter, but I'd rather like to avoid relying on it.

Reply:
I guess it's because it's a work in progress but the optimization can work on any numeric and date field ?

Reply:
@jimczi The optimization only works for the numeric long field and date field.  Date field was added as well